File: src/main/java/dev/esophose/discordbot/command/commands/misc/SayCommand.java
package dev.esophose.discordbot.command.commands.misc;

import dev.esophose.discordbot.Sparky;
import dev.esophose.discordbot.command.DiscordCommand;
import dev.esophose.discordbot.command.DiscordCommandMessage;
import discord4j.core.object.entity.channel.GuildMessageChannel;
import discord4j.core.object.util.Permission;
import discord4j.core.object.util.PermissionSet;
import java.util.Collections;
import java.util.List;

public class SayCommand extends DiscordCommand {

    public SayCommand(Sparky bot) {
        super(bot);
    }

    public void execute(DiscordCommandMessage message, String text) {
        message.delete().subscribe();
        message.getChannel()
                .cast(GuildMessageChannel.class)
                .flatMap(x -> x.createMessage(text))
                .subscribe();
    }

    @Override
    public String getName() {
        return "say";
    }

    @Override
    public List<String> getAliases() {
        return Collections.emptyList();
    }

    @Override
    public String getDescription() {
        return "Makes the bot say a message";
    }

    @Override
    public PermissionSet getRequiredBotPermissions() {
        return PermissionSet.of(Permission.SEND_MESSAGES);
    }

    @Override
    public Permission getDefaultRequiredMemberPermission() {
        return Permission.ADMINISTRATOR;
    }

}

File: src/main/kotlin/dev/esophose/discordbot/command/commands/misc/SayCommand.kt
package dev.esophose.discordbot.command.commands.misc

import dev.esophose.discordbot.command.DiscordCommand
import dev.esophose.discordbot.command.DiscordCommandMessage
import discord4j.core.`object`.entity.channel.GuildMessageChannel
import discord4j.rest.util.Permission
import discord4j.rest.util.PermissionSet

class SayCommand : DiscordCommand() {

    override val name: String
        get() = "say"

    override val aliases: List<String>
        get() = emptyList()

    override val description: String
        get() = "Makes the bot say a message"

    override val requiredBotPermissions: PermissionSet
        get() = PermissionSet.of(Permission.SEND_MESSAGES)

    override val defaultRequiredMemberPermission: Permission
        get() = Permission.ADMINISTRATOR

    fun execute(message: DiscordCommandMessage, text: String) {
        message.delete().subscribe()
        message.channel
                .cast(GuildMessageChannel::class.java)
                .flatMap { it.createMessage(text) }
                .subscribe()
    }

}
